### PR TITLE
Parallelize parts of model initialization 

### PR DESCRIPTION
**Pull request by noraabiakar**
Use the task system to parallelize parts of the model initialization phase, namely cell building,`fvm_cv_discretize` and `fvm_build_mechanism_data`. 

These changes speed up large cell group builds.

Addresses #634

### CODE REVIEW

**Comment by bcumming** [comment]
The speedup of 1.9x is with how many cores?

File: arbor/fvm_layout.cpp [hunk 1]
```diff
@@ -446,12 +447,16 @@ fvm_cv_discretization fvm_cv_discretize(const cable_cell& cell, const cable_cell
...
+    std::vector<fvm_cv_discretization> cell_disc(cells.size());
+    threading::parallel_for::apply(0, cells.size(), ctx.thread_pool.get(),
+          [&] (int i) { cell_disc[i]=fvm_cv_discretize(cells[i], global_defaults);});
 
-    for (const auto& c: cells) {
-        append(combined, fvm_cv_discretize(c, global_defaults));
+    fvm_cv_discretization combined;
+    for (auto cell_idx: count_along(cells)) {
+        append(combined, cell_disc[cell_idx]);
     }
```
Comment (halfflat) [inline arbor/fvm_layout.cpp]: Would there be any material benefit in implementing a parallel reduction?

File: arbor/fvm_layout.hpp [hunk 2]
```diff
@@ -112,7 +113,7 @@ fvm_cv_discretization& append(fvm_cv_discretization&, const fvm_cv_discretizatio
 // Construct fvm_cv_discretization from one or more cells.
 
 fvm_cv_discretization fvm_cv_discretize(const cable_cell& cell, const cable_cell_parameter_set& global_dflt);
-fvm_cv_discretization fvm_cv_discretize(const std::vector<cable_cell>& cells, const cable_cell_parameter_set& global_defaults);
+fvm_cv_discretization fvm_cv_discretize(const std::vector<cable_cell>& cells, const cable_cell_parameter_set& global_defaults, const arb::execution_context& ctx);
```
Comment (halfflat) [inline arbor/fvm_layout.hpp]: Here and below: if we give the context parameter a default value (e.g. a default-constructed `execution_context`) I believe we could avoid the modifications to the unit tests (save for any parallel tests of course!)

File: arbor/fvm_layout.cpp (hunk 1 above)
Comment (halfflat) [reply arbor/fvm_layout.cpp]: On thinking about this, I realise there can be no gain here. Sorry!

**Comment by noraabiakar** [comment]
I tested this on my laptop (4 cores; i7-7600U): Prior to these changes, building a cell_group with 10000 cells takes around 8s; 60% in `fvm_build_mechanism_data`; 25% in `fvm_cv_discretize` and 10% in building the actual cells (`get_cell_description`). After the changes, the time goes down to 4.2s. The `append` functions aren't the main remaining hotspots, for further improvement we'd need to look at the content of `fvm_build_mechanism_data` and `fvm_cv_discretize`. I've already achieved some minor improvements there for a test case with many synapses, and I'm continuing to look at other opportunities for optimization.

File: arbor/fvm_layout.hpp (hunk 2 above)
Comment (noraabiakar) [reply arbor/fvm_layout.hpp]: Okay, I will make the change :)

File: arbor/fvm_lowered_cell_impl.hpp [hunk 3]
```diff
@@ -356,15 +356,17 @@ void fvm_lowered_cell_impl<B>::initialize(
...
-        catch (util::bad_any_cast&) {
-            throw bad_cell_description(rec.get_cell_kind(gid), gid);
-        }
-    }
+    cells.resize(ncell);
+    threading::parallel_for::apply(0, gids.size(), context_.thread_pool.get(),
+           [&](cell_size_type i) {
+               auto gid = gids[i];
+               try {
+                   cells[i] = std::move(any_cast<cable_cell>(rec.get_cell_description(gid)));
```
Comment (halfflat) [inline arbor/fvm_lowered_cell_impl.hpp]: The `std::move` here is redundant, but in order to remove one stage of move-construction, one can use `any_cast<cable_cell&&>`.

Comment (noraabiakar) [reply arbor/fvm_lowered_cell_impl.hpp]: Thanks! I hadn't looked closely at what `any_cast` was doing.

**Comment by halfflat** [comment]
Note: travis failure was spurious; accepting.

**Pull request merged**